OpenRA.Mods.Common/Traits/Buildings/RepairableBuilding.cs

```diff
@@ -150,11 +153,50 @@ void ITick.Tick(Actor self)
 				// The cost is the same regardless of the amount of people repairing
 				var hpToRepair = Math.Min(Info.RepairStep, health.MaxHP - health.HP);
 
-				// Cast to long to avoid overflow when multiplying by the health
-				var cost = Math.Max(1, (int)(((long)hpToRepair * Info.RepairPercent * buildingValue) / (health.MaxHP * 100L)));
+				// Have all cost values multiplied by 100, cost tends to be miniscule. 100x comes from Info.RepairPercent
+				// Cast to long to avoid overflow
+				var cost = (int)((long)hpToRepair * Info.RepairPercent * buildingValue / health.MaxHP);
 
-				// TakeCash will return false if the player can't pay, and will stop him from contributing this Tick
-				var activePlayers = Repairers.Count(player => player.PlayerActor.Trait<PlayerResources>().TakeCash(cost, true));
+				// We are using the highest possible repair cost to find the players that
+				// can pay it as we cannot know what the actual cost will be
+				var maxMultiplier = 0;
+				for (var cnt = 0; cnt < Repairers.Count; cnt++)
+				{
+					var multiplier = Info.RepairBonuses[cnt] / (cnt + 1);
+					if (multiplier > maxMultiplier)
+						maxMultiplier = multiplier;
+				}
+
+				activePlayers = Repairers.Count(player => player.PlayerActor.Trait<PlayerResources>().CanTakeCash(Math.Max(1, (int)(builtUpCost + cost * maxMultiplier / 100))));
+
+				// Reduce or increase cost based on repair bunuses and amount people repairing
+				cost /= Info.RepairBonuses[activePlayers - 1] / 100;
+
+				// Since we use 100x values we need to wait till we reach 100 to convert them back
+				builtUpCost += cost;
+				builtUpXP += cost * Info.PlayerExperiencePercentage / 100;
+
+				if (builtUpCost >= 100)
+				{
+					var regularCost = builtUpCost / 100;
+					builtUpCost -= regularCost * 100;
+
+					foreach (var player in Repairers)
+						player.PlayerActor.Trait<PlayerResources>().TakeCash(regularCost, true);
```

Comment:
This seems fishy, (ignoring the crash above) we might be taking cash from players while `activePlayers` is 0. And then no repairing will be happening despite cash being deducted.

Reply:
The other way around is also possible: Getting repairs while not paying anything. While `builtUpCost` is less then 100 we might still be repairing, and another trait might take cash from the player meanwhile, such that no cash can be removed here for the repairs that have already been done.

Reply:
moved the check `RepairActive` check upwards